Implementation for SE-0228: Fix ExpressibleByStringInterpolation

beccadax opened the pull request:
Just a rebase of the previous PR, #18590.

This PR implements a new string interpolation API with a different, finalized ABI from what we're currently shipping. We'd like to land it before the ABI freezes; otherwise we'll have to also support the existing initializers forever.

Remaining issues to resolve:

* ~1% compiler performance regression (down from about 1.5% before). Still looking into that final percentage point.
* StringWordBuilderReservingCapacity showing ~10% slowdown. Need to investigate.
* A few other slowdowns in random benchmarks with no apparent connection to string interpolation. Need to see how much of that is noise and how much is real.
* I need to do one last sweep through it for little cleanups I might have missed.

Issues we'll wait to address:

* `Double` and `Float80` are faster to interpolate than before, but `Float` is slower. It's hard to test this in isolation while the rest of the branch is in flux, so we plan to land the change and then look at `Float` interpolation.
* On Linux only, the optimizer can't completely optimize away a string interpolation whose result is not used. We were already skipping this test on 32-bit architectures because it was broken there. Filed as [SR-9008](https://bugs.swift.org/browse/SR-9008) so we don't lose track of it.

Implements [SE-0228](https://github.com/apple/swift-evolution/blob/master/proposals/0228-fix-expressiblebystringinterpolation.md). Resolves [SR-1260](https://bugs.swift.org/browse/SR-1260), [SR-2303](https://bugs.swift.org/browse/SR-2303), and [SR-3969](https://bugs.swift.org/browse/SR-3969). Resolves rdar://43621912.

cc @milseman @ravikandhadai

beccadax committed:
Tweak TypeChecker::callWitness() signature

It doesn’t actually need its array to be mutable.

beccadax committed:
Character->String through a less circular path

beccadax committed:
Extract NSCoding unstable name diagnosis

We’ll want to add more code near this, so it’s threatening to get unweildy.

beccadax committed:
Add TapExpr

This is not used for anything yet, but it will contain the append* calls on the StringInterpolation instance.

beccadax committed:
Permit empty interpolations

As in, “foo\()baz”. These were previously hard syntax errors due to a quirk in Parser::parseList(); there’s no particular reason they should be.

beccadax committed:
Refactor getCalledValue()

To also support a not-yet-used getCalledDeclRef().

beccadax committed:
Rework ExpressibleByStringInterpolation

beccadax committed:
Rip out typeCheckExpressionShallow()

It is now dead code.

beccadax committed:
Tweak fix-it locations for throwing interpolators

beccadax committed:
Diagnose missing appendInterpolation methods

Since we can’t declare a formal appendInterpolation(…) requirement, we manually check that there is at least one method which:

* is as public as the type itself,
* returns Void or has an @discardableResult, and
* is not static.

If there isn’t, we emit an error on the conformance and a warning for each appendInterpolation method present, indicating why it is not suitable.

beccadax committed:
Warn about strange interpolations in Swift 4.2

beccadax committed:
Late-breaking string interpolation test fixes

beccadax committed:
Remove @_effects(readonly) helper functions

New string interpolation design can’t use @_effects(readonly), so these are no longer useful.

beccadax committed:
Apply feedback from @slavapestov

beccadax committed:
Simplify InterpolatedStringLiteralExpr::forEachSegment()

Removes the helper InterpolatedStringLiteralExpr::SegmentInfo struct. This also allows us to bring back the old ApplyExpr::getCalledValue() implementation.

beccadax committed:
Fix issues from rebasing

beccadax committed:
Update for DeclContext renaming

beccadax committed:
Fix method call on null object

beccadax committed:
Make both same-file extensions and default args work

Previously, the parser generated compound method names for appendInterpolation(…) methods because this helped when finding appendInterpolation methods declared in the same file. However, this turned out to prevent default arguments from working.

This commit returns it to adding base names only and instead explicitly calls loadAllMembers() on the StringInterpolation type. I’m not sure why types in other expressions don’t need this but types in these generated expressions do, but it’s much closer to the problem and doesn’t seem to have any ill effects.

beccadax committed:
Improve debuggability of lookupDirect()

This change:

1. Adds accessors for the bit attached to NominalTypeDecl::LookupTable. These let you more easily search for and break on changes; more importantly, they give this bit a name and some documentation explaining what (I think) it means.

2. Includes the “ignoreNewExtensions” parameter in the debug output from NominalTypeDecl::lookupDirect().

3. Adds a MemberLookupTable::dump() method.

beccadax committed:
Remove dead MemberLookupTable::addExtensionMembers()

These methods are never used and appear to be holdovers from an earlier implementation.

beccadax committed:
Update lazy LookupTable when extension added

Previously, nothing would guarantee that, if a nominal type `T` with lazy members had extensions adding members with name `foo`, and these members were already in `T`’s LookupTable, and a new extension to `T` added another member named `foo`, the new extension’s member would be added to the LookupTable.

This change makes it so that adding an extension clears the isLookupTablePopulated() flag, and so that when the flag is cleared on a type with lazy members, it updates all existing entries instead of just the ones present in the type itself.

Finally, it removes a workaround in string interpolation which is no longer necessary because of this change.

beccadax committed:
Fix Linux-only build failure

Apparently, the macOS STL is more forgiving about tuples vs. pairs than the Linux one.

beccadax committed:
Benchmark interpolation wtih custom types

beccadax committed:
Add missing @inlinable annotation

Probably responsible for at least part of performance gap.

beccadax committed:
Correct arguable CustomStringInterpolation benchmark flaw

As currently written, the optimizer can completely remove the involvement of CustomString in some cases. We probably don’t want that, so let’s fix it.

beccadax committed:
Test inlining an empty-string check

beccadax committed:
Inline the SmallString check for reserveCapacity()

beccadax committed:
Update for function type changes

beccadax committed:
Update sil_location test for updated mangling

beccadax committed:
Change suspect code completion test

This change tests that, in string interpolation, code completion doesn’t suggest Void functions but suggests functions returning other types.

This is a change in behavior, but I’m not convinced the old version was correct in the first place—why would you want to interpolate a function that returns Void? It’s valid but not useful, and code completion seems to try to avoid suggesting Void functions in other similar circumstances.

beccadax committed:
Use multiline string literals in examples

beccadax committed:
Remove inaccurate comments

beccadax committed:
Improve "informal requirement" comment

beccadax committed:
Test mixture of throwing and nonthrowing interpolations

beccadax committed:
Style fixes

beccadax committed:
Delete ExpressibleByStringInterpolation deprecation test

beccadax committed:
XFAIL Linux-only test failure

This already didn't work on 32-bit platforms; now it's not working on 64-bit Linux. Filed as SR-9008.

beccadax committed:
Tweak constraints to improve performance

brentdax commented:
@swift-ci please test

brentdax commented:
@swift-ci please test compiler performance

brentdax commented:
@swift-ci please smoke benchmark

eeckstein commented:
We should investigate the performance and code size regressions before we land this

beccadax committed:
Remove new string interpolation benchmarks

They will return, probably in a separate pull request; I just want to see how much they're inflating StringInterpolation.o's code size.

brentdax commented:
So StringInterpolation.o's existing benchmarks grew 5% in -O and 9.6% in -Osize; it's only the new benchmarks that made it look like it had grown 50-60%.

Still need to get a handle on the rest of the code size increase, of course. I'm guessing it's because we're inlining more—we just need to find the right balance there.

Before that, though, let's try a potential cheap fix for the FloatingPointPrinting_Float_interpolated performance regression.

brentdax commented:
@swift-ci please smoke benchmark

brentdax commented:
Did some offline benchmarking since CI was looking a little noisy. The `TextOutputStreamable` conformance does actually improve the performance of `FloatingPointPrinting_Float_interpolated`:

IMPLEMENTATION | MIN(μs) | MAX(μs) | MEAN(μs) | SD(μs) | MEDIAN(μs)
-- | -- | -- | -- | -- | --
No TextOutputStreamable | 108,746 | 114,023 | 109,509 | 571 | 109,359
description.write(to:) | 108,830 | 111,132 | 109,499 | 524 | 109,311
_writeASCII(_:) | 106,688 | 111,552 | 107,625 | 725 | 107,491

It's just that something else about new string interpolation is slower for all float types than the old version, and `TextOutputStreamable` helps `Double` and `Float80` enough to make up for it, but not `Float`.

Comparing [this branch's -emit-sil dump of `run_FloatingPointPrinting_Float_interpolated(_:)` to master's](https://gist.github.com/brentdax/3627399242f71877a19d4479d85e18fd), this branch generates 551 lines of SIL in 70 basic blocks, while master generates only 390 lines in 47 basic blocks. The broad sketches of the two functions are similar, but this branch generates a number of basic blocks that either are never generated in master or are optimized away. I'll try to clean things up tomorrow so I can get a reasonable diff and figure out what the changes represent.

brentdax commented:
A large part of the problem is [the inlining of an early exit from `append(_:)`](https://github.com/apple/swift/pull/19963/files#diff-063959f878b3333c947ff31c202d498fR989). This is intended to improve performance, and on the whole it does, but it also tends to hugely convolute control flow. For example, `run_Fibonacci(_:)` (which uses string interpolation only on a failure path inlined from `CheckResults(_:)`) has 13 basic blocks without this inlining, but 69(!) with it. It also inflates code size (changes < 1% omitted):

Benchmark | With inlining | Without inlining | Percentage Change
-- | -- | -- | --
**Regression (2)** | | |
MapReduce.o | 34,875 b | 35,627 b | 2.2%
DropWhile.o | 23,835 b | 24,251 b | 1.7%
**Improvement (74)** | | |
DeadArray.o | 3,472 b | 2,437 b | -29.8%
ByteSwap.o | 3,029 b | 2,245 b | -25.9%
Fibonacci.o | 2,869 b | 2,165 b | -24.5%
LinkedList.o | 3,013 b | 2,309 b | -23.4%
StringInterpolation.o | 16,241 b | 12,529 b | -22.9%
Exclusivity.o | 4,858 b | 3,861 b | -20.5%
SequenceAlgos.o | 31,691 b | 25,531 b | -19.4%
MonteCarloPi.o | 2,528 b | 2,149 b | -15.0%
SevenBoom.o | 2,560 b | 2,181 b | -14.8%
RangeIteration.o | 2,592 b | 2,213 b | -14.6%
PointerArithmetics.o | 2,640 b | 2,261 b | -14.4%
ProtocolDispatch2.o | 2,653 b | 2,277 b | -14.2%
BitCount.o | 2,848 b | 2,469 b | -13.3%
NSDictionaryCastToSwift.o | 3,072 b | 2,693 b | -12.3%
Integrate.o | 3,104 b | 2,725 b | -12.2%
StackPromo.o | 3,435 b | 3,035 b | -11.6%
Ackermann.o | 3,520 b | 3,141 b | -10.8%
FloatingPointPrinting.o | 7,221 b | 6,453 b | -10.6%
ArrayLiteral.o | 3,888 b | 3,509 b | -9.7%
DictionaryBridge.o | 3,995 b | 3,627 b | -9.2%
StringBuilder.o | 16,443 b | 14,939 b | -9.1%
TwoSum.o | 13,163 b | 11,963 b | -9.1%
Memset.o | 4,368 b | 3,989 b | -8.7%
RandomValues.o | 4,336 b | 3,957 b | -8.7%
XorLoop.o | 4,336 b | 3,957 b | -8.7%
OpenClose.o | 5,237 b | 4,871 b | -7.0%
Calculator.o | 5,648 b | 5,257 b | -6.9%
DriverUtils.o | 222,971 b | 208,251 b | -6.6%
HashQuadratic.o | 12,571 b | 11,803 b | -6.1%
PopFrontGeneric.o | 12,203 b | 11,499 b | -5.8%
ChainedFilterMap.o | 4,834 b | 4,562 b | -5.6%
ObjectAllocation.o | 4,552 b | 4,325 b | -5.0%
ObjectiveCNoBridgingStubs.o | 7,616 b | 7,237 b | -5.0%
DictionaryKeysContains.o | 29,403 b | 27,989 b | -4.8%
MonteCarloE.o | 6,795 b | 6,475 b | -4.7%
RC4.o | 8,139 b | 7,771 b | -4.5%
NopDeinit.o | 8,241 b | 7,888 b | -4.3%
RomanNumbers.o | 20,155 b | 19,291 b | -4.3%
DictTest2.o | 26,987 b | 25,867 b | -4.2%
DropLast.o | 33,051 b | 31,675 b | -4.2%
StrComplexWalk.o | 8,944 b | 8,568 b | -4.2%
DictionarySwap.o | 44,331 b | 42,651 b | -3.8%
Suffix.o | 35,403 b | 34,043 b | -3.8%
StrToInt.o | 10,237 b | 9,869 b | -3.6%
ArraySubscript.o | 10,763 b | 10,395 b | -3.4%
Queue.o | 21,339 b | 20,603 b | -3.4%
RangeAssignment.o | 11,323 b | 10,955 b | -3.3%
ObjectiveCBridging.o | 56,635 b | 54,891 b | -3.1%
DictionaryGroup.o | 29,307 b | 28,491 b | -2.8%
ObjectiveCBridgingStubs.o | 10,512 b | 10,215 b | -2.8%
SortLettersInPlace.o | 13,632 b | 13,258 b | -2.7%
CountAlgo.o | 26,651 b | 25,963 b | -2.6%
DictionaryBridgeToObjC.o | 10,539 b | 10,267 b | -2.6%
RangeReplaceableCollectionPlusDefault.o | 14,171 b | 13,803 b | -2.6%
Substring.o | 45,563 b | 44,379 b | -2.6%
Combos.o | 28,331 b | 27,611 b | -2.5%
PopFront.o | 11,547 b | 11,275 b | -2.4%
DictTest4.o | 36,459 b | 35,611 b | -2.3%
RGBHistogram.o | 37,003 b | 36,203 b | -2.2%
Walsh.o | 14,811 b | 14,491 b | -2.2%
Hash.o | 37,819 b | 37,019 b | -2.1%
NibbleSort.o | 17,243 b | 16,875 b | -2.1%
COWTree.o | 18,539 b | 18,171 b | -2.0%
TestsUtils.o | 21,067 b | 20,699 b | -1.7%
SortIntPyramids.o | 17,051 b | 16,779 b | -1.6%
CString.o | 24,455 b | 24,087 b | -1.5%
DictionaryCompactMapValues.o | 37,115 b | 36,555 b | -1.5%
LuhnAlgoEager.o | 24,459 b | 24,091 b | -1.5%
LuhnAlgoLazy.o | 24,459 b | 24,091 b | -1.5%
SortLargeExistentials.o | 23,787 b | 23,419 b | -1.5%
LazyFilter.o | 13,947 b | 13,755 b | -1.4%
StringRemoveDupes.o | 27,227 b | 26,859 b | -1.4%
WordCount.o | 86,539 b | 85,403 b | -1.3%
BinaryFloatingPointProperties.o | 15,755 b | 15,579 b | -1.1%
DictionarySubscriptDefault.o | 47,915 b | 47,387 b | -1.1%

Even with the costs, though, inlining this check is often profitable, and removing it will cause performance regressions. On an iMac Pro:

<details open>
  <summary>With inlining vs. without — Regression (8)</summary>
  
TEST                                                      | WITH INLINING    | WITHOUT INLINING    | DELTA  | RATIO     
---                                                       | ---    | ---    | ---    | ---       
DataReplaceLarge                                          | 19054  | 23629  | +24.0% | **0.81x (?)** 
PointerArithmetics                                        | 22377  | 26316  | +17.6% | **0.85x** 
StringEqualPointerComparison                              | 245    | 273    | +11.4% | **0.90x** 
DropWhileArrayLazy                                        | 88     | 96     | +9.1%  | **0.92x** 
PrefixWhileAnyCollectionLazy                              | 35     | 38     | +8.6%  | **0.92x** 
DropWhileAnyCollectionLazy                                | 65     | 70     | +7.7%  | **0.93x** 
DataAppendSequence                                        | 12690  | 13485  | +6.3%  | **0.94x (?)** 
ArrayAppendFromGeneric                                    | 335    | 355    | +6.0%  | **0.94x (?)** 

</details>

<details >
  <summary>With inlining vs. without — Improvement (14)</summary>
  
TEST                                                      | WITH INLINING    | WITHOUT INLINING    | DELTA  | RATIO     
---                                                       | ---    | ---    | ---    | ---       
Dictionary4                                               | 319    | 231    | -27.6% | **1.38x** 
SumUsingReduceInto                                        | 450    | 349    | -22.4% | **1.29x** 
SumUsingReduce                                            | 451    | 353    | -21.7% | **1.28x** 
DataAppendDataLargeToMedium                               | 23393  | 18523  | -20.8% | **1.26x (?)** 
Dictionary4OfObjects                                      | 383    | 311    | -18.8% | **1.23x** 
StaticArray                                               | 6      | 5      | -16.7% | **1.20x (?)** 
DataCount                                                 | 23     | 20     | -13.0% | **1.15x** 
DictionaryCompactMapValuesOfCastValue                     | 10597  | 9329   | -12.0% | **1.14x** 
DataAppendBytes                                           | 3939   | 3564   | -9.5%  | **1.11x (?)** 
StringAdder                                               | 863    | 791    | -8.3%  | **1.09x** 
FatCompactMap                                             | 876    | 810    | -7.5%  | **1.08x** 
DataAppendDataSmallToMedium                               | 4933   | 4606   | -6.6%  | **1.07x (?)** 
DataAppendDataMediumToMedium                              | 5261   | 4917   | -6.5%  | **1.07x** 
ObjectiveCBridgeStubNSDataAppend                          | 1678   | 1595   | -4.9%  | **1.05x (?)** 

</details>

The test being inlined is complicated—probably unnecessarily—and I'm hoping a simpler version will get us the best of both worlds. Barring that, perhaps the SIL optimization folks can make some suggestions. I mean, we can do better than this:

![Incredibly convoluted control flow graph, with almost all of the blocks in the failure path.](https://user-images.githubusercontent.com/931226/47468103-63983180-d7ae-11e8-988f-a922c8f7f79a.png)

beccadax committed:
Inline only part of the append(_:) length check

Gets most of the performance win without most of the code size loss.

milseman commented:
Just do the part that checks for the canonical empty string, that eliminates most of that CFG.

brentdax commented:
@milseman That’s what I left my computer benchmarking—`_isEmptySingleton` check in an inlinable method, `count == 0 && !_isNative` in a non-inlinable one.

brentdax commented:
On my own machine, with everything closed down, literally unplugged from the network, I got wildly different benchmark results for this last night vs. this morning. But I think this is at least an improvement. We'll see what CI says.

xedin commented on lib/Sema/CSGen.cpp:
```diff
@@ -3101,6 +3122,18 @@ namespace {
+
+      if (auto subExpr = expr->getSubExpr()) {
+        auto subExprType = CS.getType(subExpr);
+        CS.addConstraint(ConstraintKind::Bind, subExprType, tv, locator);
```
One small change which I think might work here - instead of biding sub-expr type to type variable, you can return `subExprType` directly and allocate type variable only if there was no sub-expression...

brentdax replied:
Lower priority than the runtime performance stuff, but I'll look at it. Thanks!

xedin commented on lib/Sema/CSGen.cpp:
```diff
@@ -1269,6 +1263,33 @@ namespace {
+          interpolationProto->lookupDirect(tc.Context.Id_StringInterpolation);
+        if (associatedTypeArray.empty()) {
+          tc.diagnose(expr->getStartLoc(), diag::interpolation_broken_proto);
+          return nullptr;
```
I think interpolation protocol lookup and field lookup could be moved to `TypeChecker`  just like we do for integers e.g. `TypeChecker::getMaxIntegerType`

brentdax replied:
Other parts of CSGen look up associated types in various ad-hoc ways like "find the one associated type in this protocol (let's hope there's just one)" or "directly call getIdentifier() with a string literal containing the associated type's name, then look it up". That's not necessarily a good thing either, though, so maybe I should just clean all of those up.

xedin commented on lib/Sema/CSGen.cpp:
```diff
@@ -1269,6 +1263,33 @@ namespace {
+        // Must be Conversion; if it's Equal, then in semi-rare cases, the 
+        // interpolation temporary variable cannot be @lvalue.
+        CS.addConstraint(ConstraintKind::Conversion, appendingExprType,
+                         interpolationTV, appendingLocator);
```
I'm really curious of what the example of behavior described in the comment might look like

brentdax replied:
There's a minor, non-ABI improvement I want to make to this code that I can't do because it triggers this behavior. Once the branch is merged, I'll redo that work and show you the problem.

brentdax commented:
@swift-ci please smoke benchmark

brentdax commented:
Gonna rebase to get floating-point string formatting improvements from @airspeedswift.

beccadax committed:
Conform float types to TextOutputStreamable

This should improve interpolation performance, but I don’t know what impact the refactoring might have on .description, so we’ll see if we keep it.

milseman commented:
reserveCapacitySlow, IIRC, can be called on lazily bridged Cocoa strings, so not native necessarily.

Otherwise, yeah, we should come up with some pattern or behavior to allow the optimizer to constant-fold all the subsequent branches.

brentdax commented:
@swift-ci please smoke benchmark

brentdax commented:
@swift-ci please smoke benchmark

beccadax committed:
Create interpolation strings with appropriate initial capacity

Should avoid relying so heavily on the optimizer realizing it can remove branches.

This had somewhat complicated performance impacts in local benchmarking; we’ll see what it does on CI.

brentdax commented:
@swift-ci please smoke benchmark

beccadax committed:
Give DefaultStringInterpolation its own append paths

This should allow us to separately control how appends are inlined for string interpolation.

brentdax commented:
@swift-ci please smoke benchmark

brentdax commented:
@swift-ci please test compiler performance

brentdax commented:
@swift-ci please test compiler performance

brentdax commented:
@swift-ci please smoke test compiler performance

brentdax commented:
@swift-ci please smoke test compiler performance

brentdax commented:
@swift-ci please smoke benchmark

brentdax commented:
@swift-ci please smoke test compiler performance

brentdax commented:
@swift-ci please smoke test compiler performance

brentdax commented:
@swift-ci please smoke benchmark

brentdax commented:
@swift-ci please test compiler performance

beccadax committed:
Skip += fast path for literal segments, too

brentdax commented:
@swift-ci please smoke benchmark

brentdax commented:
@swift-ci please smoke benchmark

brentdax commented:
@swift-ci please smoke benchmark

brentdax commented:
So inlining the empty-string check is a bad idea. Let's see if skipping it during interpolations will help; if so, we may be able to look into an optimizer pass (in the future) to handle the cases where we don't need to do a real append.

brentdax commented:
@swift-ci please smoke benchmark

brentdax commented:
Fascinating. [That last set of benchmark results](https://github.com/apple/swift/pull/19963#issuecomment-434148535) was for [this now-reverted change](https://github.com/brentdax/swift/compare/a91cbd63daba54dbc5714b58ad963a99d9d4d912...brentdax:interpolation-rework-reverted). It seems like inlining the empty-string check is profitable, but only if it is *not* done for interpolations, just for other string concatenations. I've rolled back to the implementation which benchmarked better.

cc @milseman

brentdax commented:
@swift-ci please smoke benchmark

brentdax commented:
Resetting an alternate implementation of `_StringGuts.init(_initialCapacity:)` which would basically hide the `init(_large:)` call from interpolating code, and thus prevent it from doing some specialization I thought might not be a great idea. (5c0f3fd152f29defcdd4c49aed356cd11c7d3a3e)

brentdax commented:
@swift-ci please benchmark

brentdax commented:
I think I didn't roll back far enough there.

brentdax commented:
@swift-ci please smoke benchmark

brentdax commented:
The over-inlining issues which caused large, widespread regressions are at this point pretty much resolved, so some specific issues are now coming into focus.

(A little bit of over-inlining may have snuck back in today; I'll be looking at that in the morning. I'm particularly suspicious of `StringBuilder.o`, `ErrorHandling.o`, `libswiftsimd.dylib`, and `AnyHashableWithAClass`, which all went away in [this benchmark run](https://github.com/apple/swift/pull/19963#issuecomment-434125342) but came back later. I may have reverted a little too much when I undid an experiment later.)

# Remaining issues

## Formatting "small" floating-point numbers is slow

* 20.9% regression in `FloatingPointPrinting_Float_interpolated` -O performance
* 21.6% regression in `FloatingPointPrinting_Float_interpolated` -Osize performance

The root cause of this performance reduction is that the new implementation is much faster for floats whose string representations are large enough to require a heap allocation, but modestly slower for floats whose string representations will fit into a small string. Virtually all `Float` values will fit into a small string, while most `Double` and `Float80` values will not (less than 1% of the values we benchmark with), so the change looks good for `Double` and `Float80` but bad for `Float`. The `Double`/`Float80` picture probably looks a little rosier than it really is because in practice, values with short representations (like "2" or "108" or "0.125") are disproportionately common, but it is not entirely a benchmark artifact.

<details>
<summary>I tried three alternate implementations, but the only one that got more than a 1.5% improvement for `Float` regressed the other types by nearly 30%.</summary>

* [interpolation-rework-float-simple](https://github.com/brentdax/swift/commit/f57638414fe08743ac9f3f23c486f0e02dbbb254): Basically returns to the implementation inherited from the old design: The formatting function returns a `String` and we append it to the `String` we're building. At -O, this improves`Float` performance by 3.6%, but regresses `Double` and `Float80` by 29% and 26% respectively.

* [interpolation-rework-float-inout](https://github.com/brentdax/swift/commit/87921d861f7128d75b93ec66101a0ec027a561db): Passes the result buffer in as an inout argument to the formatting function, rather than returning it from the formatting function. At -O, this regresses all three types' interpolation performance by <2%.

* [interpolation-rework-float-with](https://github.com/brentdax/swift/commit/8f7dba8a79ca9ad81f48d4e4d37223c710df8d26): Passes a closure to the formatting function which uses the buffer. At -O, this appeared to improve all three types, but only by <1.5%.
</details>

There might be other approaches which would help; we might also consider a hybrid design which uses different strategies for `Float`s and larger numbers, or for numbers which will fit into a small string and numbers which won't. The good news is, the specific implementation of floating-point formatting is not part of the ABI; only the fact that they can be formatted through `TextOutputStreamable` is. The standard library and performance teams can look at this later.

I had hoped for a simple solution, but ultimately I don't think this one benchmark should be a showstopper, especially when we can probably improve it in the future.

## Benchmark artifact or overly conservative estimate

* 19.1% regression in `StringInterpolation` -O performance
* 32.3% regression in `StringInterpolation` -Osize performance

As I've said before, this is probably a benchmark artifact—`StringInterpolation` used to resize its internal buffer twice; now it resizes it three times. A slightly different pattern of segment sizes would give a different result.

It's possible, however, that segment sizes like these are pretty common and code in the wild will see similar issues. If so, this could be fixed by substantially increasing the `capacityPerInterpolation` constant in StringInterpolation.swift. It is currently 2, meaning two extra characters of storage are pre-allocated per interpolated value in the string. If this estimate is too conservative, we can increase it or substitute a more sophisticated estimate (for instance, always trying to use a small string if there's any possibility we can fit inside one, or using a large estimate but capping the amount of space we will potentially waste). The old calculation will still be inlined into old code, but new code can use a different calculation without breaking the ABI.

## Dead code elimination can't reliably eliminate interpolations

* Failure of `SILOptimizer/unused_container.swift` test on Linux (currently XFAILed)
* 16.7% regression in `libswiftSwiftReflectionTest.dylib` code size
* 3.6% regression in -O code size of `DeadArray.o`
* 12.2% regression in -Osize code size of `DeadArray.o`

Dead code elimination is not as good at understanding this new design as it was with the old one. The optimizer doesn't know that the calls to mutating `appendLiteral(_:)` and `appendInterpolation(_:)` methods only mutate `self`, so it doesn't know that they're safe to eliminate. I'm hoping that @eeckstein or @aschwaighofer might know of a quick solution, but even if there's nothing ready-made, I can't see why it wouldn't be solvable without an ABI change.

## Poorly understood

* 15.5% regression in `RandomDoubleLCG` -O performance
* Regressions in -Osize performance not otherwise mentioned

These have no obvious cause I can discern; there's no smoking gun I can find in the SIL. It's possible that resolving one of the known issues will help with them, or that profiling will reveal something.

## Not seriously investigated

* -Onone performance regressions

The set of regressions is quite different from optimized compiles, none of them are mind-bogglingly large (particularly in the earlier run I mentioned in the first section), and they seem to have responded a lot to earlier rounds of tuning.

# Plan of action

I'm going to look into the dead code elimination issue tomorrow, but even if there's no quick fix, I think we should merge this soon (perhaps tomorrow). UTF-8 strings may change some of these performance tradeoffs dramatically, and the biggest regressions are either gone or at least don't look like they'll be frozen in the ABI. Merging this to master will let the entire team independently chase down the remaining issues and allow @milseman to stop worrying about giant merge conflicts.

eeckstein commented:
I'll take a look at the dead-code-elimination problem. This should not block you from merging.

brentdax commented:
@swift-ci please smoke benchmark

brentdax commented:
@swift-ci please smoke benchmark

brentdax commented:
I…guess that's a Halloween smoke benchmark full of spooky zombie regressions?

brentdax commented:
@swift-ci please benchmark

brentdax commented:
I have a working, but very hacky, set of SILOptimizer improvements to properly eliminate dead code with this design. They need a little more time to bake, but I want to benchmark them and confirm they'll help with the remaining regressions before I merge. So I will:

1. Push the WIP optimizer improvements.
2. Benchmark.
3. Revert the WIP optimizer improvements (I'll put them on a branch).
4. Run a final round of tests and source compatibility checks to make sure I don't break anything.
5. Merge.

That will land the portion of this project which directly changes the ABI and let everyone (myself included) stop worrying about merge conflicts.

brentdax commented:
@swift-ci please benchmark

brentdax commented:
After reverting, the optimization WIP commit will still be available on [brentdax/swift's interpolation-rework-silopt branch](
https://github.com/brentdax/swift/tree/interpolation-rework-silopt).

brentdax commented:
Interesting. The WIP dead code elimination excludes `DefaultStringInterpolation.appendLiteral(_:)` and `DefaultStringInterpolation.appendInterpolation(_:)` from early inlining; that's probably why it causes performance regressions.

But even with that effect, proper dead code elimination swings most of the suspected code size issues from regressions to improvements:

File | Without DCE | With DCE
-- | -- | --
libswiftSwiftReflectionTest.dylib | +16.7% | -8.3%
DeadArray.o at -O | +3.6% | -14.7%
DeadArray.o at -Osize | +12.2% | -7.1%

So these three regressions—which are among the most serious remaining ones—do come from dead code elimination not working, and we should expect them to improve in the future. A handful of other benchmarks did also improve performance with the WIP despite the inlining headwinds.

Unfortunately, this DCE implementation obviously isn't there yet. But it's the right direction to look in.

beccadax committed:
Confess my (current) source and ABI stability sins
